### DISCUSSION TITLE
Adding Bulk Data into Pocketbase Database

### DISCUSSION THREAD

**Discussion by gaganbiswas** [pocketbase/pocketbase#4230] [Q&A] [answered]
```ts
  const response = await Promise.all(
    jsonData.map(async (data) => {
      const res = await fetch(process.env.POCKETBASE_URL!, {
        method: "POST",
        body: JSON.stringify(data),
      });

      const resData = await res.json();
      if (resData.id) {
        return {
          id: resData.id,
          status: 1,
          message: { title: "", description: "" },
        };
      } else {
        return {
          id: "",
          status: 0,
          message: { title: "Server Error", description: resData.message },
        };
      }
    })
  );
```

I am getting an `Method not allowed` error on this. Can you tell me if I am doing something wrong? I need to bulk upload these data into the database. The data is an array of object.

**Comment by ganigeorgiev** [comment, marked as answer]
What is `process.env.POCKETBASE_URL`?

If you are trying to submit a POST request to the records create web API then the url should be something like: `BASE_URL/api/collections/:collectionIdOrName/records`.

Alternatively, you could use the [JS SDK](https://github.com/pocketbase/js-sdk), although for this particular case you'll have to disable the [autocancellation](https://github.com/pocketbase/js-sdk?tab=readme-ov-file#auto-cancellation) by either specifying a unique `requestKey` for each request or just set it to `null`.

**Reply by gaganbiswas** [reply]
Okay. I got where I was wrong. 

Also, Is this a good approach? Like we still don't have an bulk upload API so will this approach be good for say like an array of data with length of < 500 ?

**Reply by ganigeorgiev** [reply]
Yes, it should be fine.

If you have to perform the above action frequently it may better to create a custom route with Go or JS and process the data on the server-side.

**Reply by ganigeorgiev** [reply]
_Side-note: instead of `Promise.all` you can also try with [`Promise.allSettled`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled) which will return something similar to your resData checks (edit: at least when the JS SDK is used because on non 20x response we throw an error)._

**Reply by gaganbiswas** [reply]
Thank you for the insight will be doing that then.

**Reply by gaganbiswas** [reply]
```js
routerAdd("POST", "/batch-write", (c) => {
  const info = $apis.requestInfo(c);
  const { data, collectionName } = info.data;

  const collection = $app.dao().findCollectionByNameOrId(collectionName);

  $app.dao().runInTransaction((txDao) => {
    for (let item of data) {
      const record = new Record(collection);

      Object.entries(item).forEach(([key, value]) => {
        record.set(key, value);
      });

      txDao.saveRecord(record);
    }
  });

  return c.json(200, { status: 1 });
});
```

I wrote this simple custom js hook inside pb_hooks, and I have some questions :
1. Will this work as intended for bulk uploads into a particular collection?
2. How shall I check the create rule? I am sending the Token as header authorization.
3. How am I supposed to call this path from my frontend/server? Will it be like --
```js
await fetch(<POCKETBASE_URL>/api/batch-write, {
  method: 'post', 
  body: JSON.stringify({data, collectionName}), 
  header: {authorization: token}
});
```

**Reply by ganigeorgiev** [reply]
1. The above should work correctly but note that it won't perform validations for the loaded data. If you need validations you'll have to use the `RecordUpsetForm`. 
2. This could be done multiple ways but the easiest would be to use the [`$app.dao().canAccessRecord`](https://pocketbase.io/docs/js-records/#check-if-record-can-be-accessed) helper.
3. You may need to add also `"content-type": "application/json"` header. For sending requests to custom routes with the JS SDK you can use the [`send()` method](https://pocketbase.io/docs/js-routing/#sending-request-to-custom-routes-using-the-sdks).

A slightly modified version of your code sample could be (_I haven't tested it_):

```js
routerAdd("POST", "/batch-write", (c) => {
  const info = $apis.requestInfo(c);
  const { data, collectionName } = info.data;

  const collection = $app.dao().findCollectionByNameOrId(collectionName);

  $app.dao().runInTransaction((txDao) => {
    for (let item of data) {
      const record = new Record(collection);

      const form = new RecordUpsertForm($app, record)

      form.setDao(txDao) // use the transaction dao for the db operations

      form.loadData(item)

      form.submit();

      // verify that the inserted record can be accessed by the current request client, otherwise rollback
      // (replace createRule with the rule you want to check)
      if (!txDao.canAccessRecord(record, info, collection.createRule)) {
        throw new ForbiddenError("The operation is not allowed...");
      }
    }
  });

  return c.json(200, { status: 1 });
});
```

But keep in mind that allowing the execution of unrestricted long running read/write transactions could affect the performance of your application since it requires db lock (I'd suggest as a minimum to add restriction on the max number of data items).

**Reply by ganigeorgiev** [reply]
_Also keep in mind that if your have `@request.data.*` rules note that `@request.data` in this case will refer to the submitted body and not the individual data of each record. If you want to be able to use `@request.data.*` then you'll need to create a copy of `info.data` and reasign it to the `item` in each data iteration._

**Reply by gaganbiswas** [reply]
Thanks a lot for clarifying everything. I will implement this and give you a feedback on the same. I am limiting my data to a max of 100 as of now.

**Reply by gaganbiswas** [reply]
I tested it and everything is working smoothly. Just a small query, I cannot set custom id, as the id field has some constraints attached like '/' cannot be used and the length must be exactly 15 chars long etc. Is there any way I can bypass this as of now?

**Reply by ganigeorgiev** [reply]
Without details why you need it I can't recommend it as it could have some unintended side-effects, but if you really really need to set a custom id you can provide an interceptor function to the submit call and try to replace the record id from there (it is executed after the validation and before the save db call):
```js
form.submit((next) => {
  return (m) => {
      m.setId("your_custom_id")
      next(m);
  }
});
```

If you need further help, please consider opening a new discussion and providing more details about your use case.

**Reply by gaganbiswas** [reply]
Ok thanks for the information. I think for now I will let it be as it is.
